---
title: "Untitled"
author: "Nick MacKnight"
date: "4/23/2024"
output: html_document
---

## Prep

The purpose of this script is to assemble a Master file that all subset files are based on for consistency. 
This is where most of the customized work will happen. 

# Project Decision Notes.
There are samples that do not have metadata recorded for all measurements. Should they be removed for consistency or retained and considered for the metadata they do possess.

For example, 

1.) *Missing WGS Genotype ID and Algal symbiont relative abundance data. But these putative genotypes possess count data from 16s and phenotypic susceptibility categories. 

GB 11
M15
M9
OF331
OF626
OF655
OF685
S564


2.) ApriGenotypeSymbiont (Michael's WGS Genotype IDs and algal symbiont relative abundances) possess 6 genotypes that do not exist in Sara's phenotypic susceptibility metadata:

"WGS-8"   "WGS-9"   "WGS-128" "WGS-129" "WGS-131" "WGS-144"

# Libraries
```{r libraries}
library(dplyr)
```

# Set working directory

```{r working directory}
setwd("~/Desktop/Ofav SCTLD/Raw Data/Resequenced Fastq/Ofav SCTLD By EXP")
```

# Products from Qiime2
```{r Products from Qiime2}
counts <- read.csv("../table/feature-table.t.csv")
metadata <- read.csv("../Merged-metadata.csv")
taxonomy <- read.csv("../taxonomy/taxonomy.csv")
```

```{r, echo = FALSE}
head(taxonomy)
```

```{r simplify taxonomy}
# Simplify Taxon ID. 
taxonomy <- taxonomy %>%
  mutate(SimplifiedTaxon = case_when(
    Confidence >= 1.00 ~ ifelse(grepl("s__", Taxon), sub(".*s__([^;]+).*", "s__\\1", Taxon), sub(".*g__([^;]+).*", "g__\\1", Taxon)),
    Confidence > 0.97 ~ ifelse(grepl("s__", Taxon), sub(".*s__([^;]+).*", "s__\\1", Taxon), sub(".*g__([^;]+).*", "g__\\1", Taxon)),
    Confidence > 0.95 ~ ifelse(grepl("g__", Taxon), sub(".*g__([^;]+).*", "g__\\1", Taxon), sub(".*f__([^;]+).*", "f__\\1", Taxon)),
    Confidence > 0.90 ~ sub(".*f__([^;]+).*", "f__\\1", Taxon),
    Confidence > 0.85 ~ sub(".*o__([^;]+).*", "o__\\1", Taxon),
    Confidence > 0.80 ~ sub(".*c__([^;]+).*", "c__\\1", Taxon),
    Confidence > 0.77 ~ sub(".*p__([^;]+).*", "p__\\1", Taxon),
    TRUE ~ sub(".*d__([^;]+).*", "d__\\1", Taxon)
  ))

# Replace/Remove s__, g__, and so on. 
taxonomy <- taxonomy %>%
  mutate(SimplifiedTaxon = gsub("^[a-z]__", "", SimplifiedTaxon),  # Remove prefixes
         SimplifiedTaxon = case_when(
           grepl("^s__", SimplifiedTaxon) ~ paste0(SimplifiedTaxon, " sp."),  # Add " sp." for s__
           grepl("^g__", SimplifiedTaxon) ~ paste0(SimplifiedTaxon, " spp."),  # Add " spp." for g__
           TRUE ~ SimplifiedTaxon  # Keep others as is
         ))
# Creating a new column of just the Order taxonomic level - nice for stacked percent barplots
taxonomy <- taxonomy %>%
  mutate(Order = sub(".*o__([^;]+);.*", "\\1", Taxon))

taxonomy$species <- taxonomy$ASV # Create a new column "species" with rownames

# concatenate the "Order" column to the "species" column. This is so that we can analyze by ASV but have a relevant Order name to evaluate that ASV when looking at a graph. 
taxonomy$species <- paste(taxonomy$Order,taxonomy$species, sep = "_")


head(taxonomy)

#to put the taxonomy in the right format
taxonomy_ps <- taxonomy
rownames(taxonomy_ps) <- taxonomy_ps[,1]
taxonomy_ps <- taxonomy_ps[,-1]
```



## Organize Metadata File
```{r}
# Starting with our original metadata
metadata

# Reading in Algal Symbiont Metadata
AprilGenotypeSymbiont <- read.csv("AprilGenotypeSymbiont.csv")

# Reading in whole genome genotype assignment and Disease Phenotype Susceptibility Groupings for genotypes
GenotypeSusceptibility <- read.csv("GenotypeSusceptibility.csv")

# I removed the following Genotypes from AprilGenotypeSymbiont and GenotypeSusceptibility because they do not have genotype IDs based on michaels work, or Symbiont percentages, but they did have susceptibility groupings. 
# List of names to remove
# names_to_remove <- c("OF331", "OF685", "OF626", "S564", "OF655", "GB 11", "M15", "M9")

# Remove rows with specified names in the #DEP.Genotype column
# AprilGenotypeSymbiont <- subset(AprilGenotypeSymbiont, !(DEP.Genotype %in% names_to_remove))
# GenotypeSusceptibility <- subset(GenotypeSusceptibility, !(DEP.Genotype %in% names_to_remove))

#Now its time to ensure the formatting is the same among our three putative metadata files so that they can be combined into one.

head(metadata)
```
```{r}
dim(AprilGenotypeSymbiont) # 180 x15
head(AprilGenotypeSymbiont)
```
```{r}
dim(GenotypeSusceptibility) # 146 x12
head(GenotypeSusceptibility)
```

```{r}
# There is unneccessary information in AprilGenotypeSymbiont
# Define the column name to filter by
keep <- c("DEP.Genotype","Sample.ID","Putative.Genotype.ID","Symbiodinium","Breviolum","Cladocopium","Durusdinium") 

# Subset the data frame to include only the specified columns
AprilGenotypeSymbiont <- AprilGenotypeSymbiont[, keep, drop = FALSE]

Genotype_Symbiont_Susceptibility <- merge(AprilGenotypeSymbiont,GenotypeSusceptibility, by="DEP.Genotype", all=T)
head(Genotype_Symbiont_Susceptibility)
```

```{r}
# Make sure the Genotypes exist in both metadata files before merging and accidently leaving some out. This is because their are minor formatting differences between putative genotype id's. 

# Find strings in AprilGenotypeSymbiont that are not in metadata
unique_strings_April <- setdiff(AprilGenotypeSymbiont$DEP.Genotype, GenotypeSusceptibility$DEP.Genotype)
print(unique_strings_April)

unique_strings_Suscep <- setdiff(GenotypeSusceptibility$DEP.Genotype, AprilGenotypeSymbiont$DEP.Genotype)
print(unique_strings_Suscep)

# Find strings in AprilGenotypeSymbiont that are not in metadata
unique_strings_April <- setdiff(Genotype_Symbiont_Susceptibility$Putative.Genotype.ID, metadata$Genotype)
print(unique_strings_April)

# Find strings in metadata that are not in AprilGenotypeSymbiont
unique_strings_metadata <- setdiff(metadata$Genotype, Genotype_Symbiont_Susceptibility$Putative.Genotype.ID)
print(unique_strings_metadata)

# This merged_metadata has not removed any samples that have blank metadata information for some measurements. 
merged_metadata <- merge(Genotype_Symbiont_Susceptibility, metadata, by.x = "Putative.Genotype.ID", by.y = "Genotype", all = T)


# Find strings in counts that are not in merged_metadata
unique_strings_counts <- setdiff(counts$SampleID, merged_metadata$SampleID)
print(unique_strings_counts)

unique_strings_mmeta <- setdiff(merged_metadata$SampleID, counts$SampleID)
print(unique_strings_mmeta) # These samples were removed during step 7 of my qiime2 pipeline where I removed any sample that did not have a timepoint (blanks, bacterial standards, and transmission frags) as these are relevant for quality contorl and not for most statistical analysis. I see some samples were also removed, they may not have had a documented timepoint. 


counts_metadata <- merge(merged_metadata, counts, by="SampleID")
```



```{r}
#n of metadata categories

unique_counts <- lapply(counts_metadata[,1:29], function(column) table(column))
print(unique_counts)


# Function to generate a summary of unique values and their counts
get_unique_counts <- function(column) {
  unique_vals <- unique(column)
  counts <- table(column)
  return(data.frame(Unique_Value = unique_vals, Count = counts[match(unique_vals, names(counts))]))
}

# Apply the function to each column of the dataset
unique_counts <- lapply(counts_metadata[,1:29], get_unique_counts)


# Print the results
for (i in 1:length(unique_counts)) {
  cat("Column:", names(unique_counts[i]), "\n")
  print(unique_counts[[i]])
  cat("\n")
}


# Combine data frames by row-binding them
unique_counts_export <- do.call(rbind, unique_counts)

getwd()
# Export the combined data frame to a CSV file
write.csv(unique_counts_export, "Metadata_Categorical_N.csv")
```




```{r}
#rarefaction - All Samples
library(vegan)
library(ggthemes)
D.spec <- specnumber(counts_metadata[,30:length(counts_metadata)])
D.specrare <- min(rowSums(counts_metadata[,30:length(counts_metadata)]))
Srare <- rarefy(counts_metadata[,30:length(counts_metadata)], D.specrare)
plot(D.spec,Srare, xlab = "Observed No. of Species", ylab="Rarefied No. of Species")
abline(0,1)
rarecurve(counts_metadata[,30:length(counts_metadata)], step = 20, sample=D.specrare, col = c("darkseagreen","cornflowerblue", "darksalmon"), cex=0.6, xlim=c(0,20000)) # limits to 20k reads. 
#rarecurve(counts_metadata[,30:length(counts_metadata)], step = 20, sample=D.specrare, col = c("darkseagreen","cornflowerblue", "darksalmon"), cex=0.6)
#pal <- colorblind_pal(counts_metadata[,30:length(counts_metadata)], Coral.Species)
```

```{r}
# Find the minimum row sum
subset_rows <- counts_metadata[,30:length(counts_metadata)]
row_sums <- rowSums(subset_rows)
minimum_row_sum <- min(row_sums)

row_sums <- as.numeric(row_sums)

h <- hist(row_sums,ylim=c(0,2000), breaks=20)
text(h$mids,h$counts,labels=h$counts, adj=c(0.5, -0.5))

# I want to see which samples have low reads
row_sums_annot <- data.frame(SampleID = counts_metadata[,1], RowSums = row_sums)
row_sums_annot <- row_sums_annot[order(row_sums_annot$RowSums), ]
row_sums_annot

# Filter out Low Read Samples To make the most of the dataset, 5141 was selected is the 1st quartile read depth. determined by qiime tools view table-filtered.qzv
#Retaining Only samples above 5141 read depth
subset <- (row_sums_annot[,2] > 5141)
subset_row_sums_annot <- row_sums_annot[(row_sums_annot[,2]>5141),]
dim(subset_row_sums_annot) #1279 samples
dim(row_sums_annot)  #1708 samples (unfiltered read depth)

minimum_row_sum <- min(subset_row_sums_annot[,2])

counts_metadata_sub <- merge(subset_row_sums_annot, counts_metadata, by = "SampleID")
counts_metadata_sub <- counts_metadata_sub[,-2]


```


```{r}
# All Samples

#Remove duplicate names in counts_metadata_sub - Duplicates were created because when genotypes were putative, symbiont and susceptibility work was conducted. Now, genotypes are confirmed via WGS. this means there are some putative genotypes with not exactly the same symbiont or susceptibility data even though multiple putative genotypes are now the same wgs genotype. This creates duplicates. I will keep the first mention of a duplicated WGS genotype. 

duplicates <- counts_metadata_sub$SampleID[duplicated(counts_metadata_sub$SampleID)] # 108 duplicates
counts_metadata_sub <- counts_metadata_sub[!duplicated(counts_metadata_sub$SampleID), ]

library(tidyverse)
#Concatenate the pre-existing ASV_merged name to begin with the wgsGenotype ID, followed by a "_".
counts_metadata_sub$DEPSampleID <- paste(counts_metadata_sub$DEP.Genotype, counts_metadata_sub$SampleID, sep = "_")
counts_metadata_sub <- counts_metadata_sub %>% relocate(DEPSampleID, .before = SampleID)

write.csv(counts_metadata_sub, file="counts_metadata_sub.csv")

colnames(counts_metadata_sub)[31:ncol(counts_metadata_sub)] <- gsub("^X", "", colnames(counts_metadata_sub)[31:ncol(counts_metadata_sub)])
counts.t <- t(counts_metadata_sub[,31:length(counts_metadata_sub)])
colnames(counts.t) <- t(counts_metadata_sub[,1])
counts.t <- as.data.frame(counts.t)


metadata <- counts_metadata_sub[,1:30]
#metadata %>% column_to_rownames(var = "SampleID")-> meta #already set to rowname
# rownames(metadata) <- metadata[,1]
# metadata <- metadata[,-1]


# mmetadata <- merged_data[,1:32] #new (4.8.24) merged_data
# metadata <- read.csv("/Users/nicholas.macknight/Desktop/Ofav SCTLD/Raw Data/Resequenced Fastq/Merged-metadata.csv")
taxonomy <- read.table(file = "/Users/nicholas.macknight/Desktop/Ofav SCTLD/Raw Data/Resequenced Fastq/taxonomy/taxonomy.txt",header = T,fill = T,row.names = 1) %>%
  mutate(across(.cols = everything(), ~str_replace(., "[a-z]__", ""))) %>% #replace any lowercase letter across all columns with an empty string
  mutate(across(.cols = everything(), ~str_replace(., ";", ""))) # replace a semicolon found across all columns with an empty string


#Remove duplicate names in mmetadata
# duplicates <- metadata$DEPSampleID[duplicated(metadata$DEPSampleID)]
# mmetadata <- mmetadata[!duplicated(mmetadata$DEPSampleID), ]

# Convert mmetadata to dataframe if it's not already
metadata <- as.data.frame(metadata)

# Set the DEPSampleID column as row names
rownames(metadata) <-NULL


metadata %>% column_to_rownames(var = "DEPSampleID")-> meta
# mmetadata %>% column_to_rownames(var = "SampleID")-> mmeta


# Define the order of levels for the "timepoint" column
timepoint_levels <- c("Pre.Exposure", "Early.Exposure", "Initial.Disease", "Final", "Main.Frag")

# Define the order of levels for the "Treatment" column
treatment_levels <- c("C", "D")

# Define the order of levels for the "Susceptibility Cluster" column
unique(metadata$RGroup.allclustered)
RGroup.allclustered_levels <- c("Highly Susceptible","Susceptible","Intermediate","Resistant")

# Define the order of levels for the "5 Susceptibility Cluster" column
unique(metadata$RGroup.all.genotypes.but.5.clusters)
RGroup.all.genotypes.but.5.clusters_levels <- c("Highly Susceptible","Susceptible","Intermediate","Resistant","1/1 Resistant")

# Reorder levels of timepoint and Treatment
metadata$timepoint <- factor(metadata$timepoint, levels = timepoint_levels)
metadata$Treatment <- factor(metadata$Treatment, levels = treatment_levels)
metadata$RGroup.allclustered <- factor(metadata$RGroup.allclustered, levels = RGroup.allclustered_levels)
metadata$RGroup.all.genotypes.but.5.clusters <- factor(metadata$RGroup.all.genotypes.but.5.clusters, levels = RGroup.all.genotypes.but.5.clusters_levels)

levels(metadata$RGroup.all.genotypes.but.5.clusters)

library(phyloseq)
otu_table <- otu_table(as.matrix(counts.t), taxa_are_rows = T)
samples <- sample_data(meta)
TAX <- tax_table(as.matrix(taxonomy))

# These files need to align in the phyloseq function. Here are the dims of a working ps file. The colnames of counts.t need to be the rownames of meta. 
dim(counts.t) # 321 1243
dim(meta) # 1243   29
dim(taxonomy) # 60123     8

ps <- phyloseq(otu_table, TAX, samples)
ps
```
